README.md
[![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/saengate/wagtail-compose)](https://github.com/saengate/wagtail-compose/releases/latest)
[![GitHub](https://img.shields.io/github/license/saengate/wagtail-compose)](LICENSE)
[![GitHub contributors](https://img.shields.io/github/contributors/saengate/wagtail-compose)](https://github.com/saengate/wagtail-compose/graphs/contributors)
[![Build Status](https://travis-ci.org/saengate/wagtail-compose.svg?branch=master)](https://travis-ci.org/saengate/wagtail-compose)

# BLOG (extendido de DJFULLAPP)

## README BLOG
Usa [wagtail](https://wagtail.io/)
https://docs.wagtail.io/en/stable/advanced_topics/add_to_django_project.html

## Descripción

Crear una blog personal usando `wagtail`.


# README Extendido de DJFULLAPP

### Imagen de Django 3.1/Vue CLI v4.4 con Nginx (Supervisor-Daphne), PostgreSQL y Apache Airflow.

## Descripción

Crear una imagen de docker con roles ansible que sirve como template para otras aplicaciones.
Al usar ansible se pueden modificar las contraseñas importantes y las configuraciones.
Solo debe tenerse cuidado en los contenedores si se cambian los puertos.

### Dependencias

* [Docker](https://docs.docker.com/engine/install/)
* [Docker-composer](https://docs.docker.com/compose/install/)
* [Ansible](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html)

### ¿Cómo instalar?

Debes tener instalado docker y ansible en tu equipo, se dejan arriba los links para su instalación.
Instalar Ansible sera necesario para modificar las contraseñas encryptadas.

En este momento es recomendable descargar directo desde el [repositorio GitHub](https://github.com/saengate/djfullapp), sin embargo dejo documentado las instrucciones para montar las imágenes con docker-compose:

![#f03c15](https://via.placeholder.com/15/f03c15/000000?text=+) **Deja el volumen comentado la primera vez que levantes el proyecto y luego copia el contenido desde el contenedor a tu carpeta de desarrollo**

```sh
docker-compose up --build
docker cp djfullapp:/tmp/django ./django
docker cp djfullapp-vue:/app/djfullapp-vue ./vue
docker cp {contenedor}:{/tmp/ansible} ${PWD}/{nombre-carpeta-destino}  # Que coincida con el volumen de cada contenedor
docker-compose down
```

Descomenta las lineas de volumen para enlazar el contenedor con tu desarrollo y vuelve a levantar los contenedores
```sh
docker-compose up
```

### Librerias de terceros

[django-easy-audit](https://github.com/soynatan/django-easy-audit) Es una aplicación de auditoria de fácil uso que he incluido en este projecto. De momento, estará en evaluación y si cumple con las espectativas se volverá parte del proyecto. Todas las instrucciones de uso se encuentran en la descripción de dicho proyecto (Realmente parece que no requiere de ninguna interacción en los modelos, así que al ser prácticamente innecesaria la interacción del desarrollador para su ejecución se evita que por olvido no se monitoree algún modelo, simplemente genial).


## Desarrollo

Te recomiendo usar alias en tu perfil de usuario para que te sea sencillo usar los scripts que facilita el proyecto
por ejemplo, esto puedes hacerlo buscando en tu $HOME el archivo `.bashrc` o `.zshrc`, entre otros, si no conoces de ellos
busca los archivos que terminan en `rc` dentro del $HOME o investiga un poco al respecto sobre como agregar las variables
de entorno en tu OS.

```sh
export PROJECT="/dir/path/djfullapp"
alias cds="$PROJECT"
alias cmd="$PROJECT/cmd"
alias cmdp="$PROJECT/django/cmdp"
alias cmdn="$PROJECT/neo4j/cmdn"
alias cmdb="$PROJECT/postgres/cmdb"
alias cmdv="$PROJECT/vue/cmdv"
```

Levantar el proyecto
```sh
docker-compose up --build -d | cmd -db
```

Se agrega el comando "cmd" para facilitar el uso del proyecto y su interacción con docker-compose y vault-ansible:
```sh
./cmd -h | cmd -h
```
```sh
-h | * | --help   muestran los comandos disponibles

-D   | --daemon             inciar el proyecto en background            (docker-compose up -d)
-DB  | --daemon_build       construye e incia el proyecto en background (docker-compose up --build -d)
-d   | --start              inciar el proyecto con verbose              (docker-compose up)
-db  | --start_build        construye e incia el proyecto con verbose   (docker-compose up --build)
-dl  | --docker_list        lista los contenedores por Nombre, Estado y (docker ps)
                            Puestos

-pm  | --proyect_migrate    ejecuta las migraciones en el proyecto      (docker exec -it...django-migrate)
-ca  | --create-admin       crea el usuario administrador por defecto   (docker exec -it...createsuperuser)
                            de la aplicación

-tp  | --tests_project      entra al contenedor del proyecto y ejecuta  (docker exec -it.../manage.py test)
                            los tests de Python
-tn  | --tests_npm          entra al contenedor de vue y ejecuta        (docker exec -it...npm run test)
                            los tests npm

-sp  | --shell_project      accede al contenedor del proyecto           (docker exec -it)

-s   | --stop               detiene los contenedores                    (docker-compose down)

-lp  | --local_persist      Para desarrollo inicia un contenedor docker (docker run local-persist)
                            con local persistencia
```

### Librerias (pip requirements.txt y poetry)

Este proyecto no usa directamente requirements para instalar el proyecto, en su lugar usa [poetry](https://pypi.org/project/poetry/) que generara el archivo durante la creación del contenedor y maneja efectivamente las versiones de las librerias. Se usa este metodo para aprovechar las ventajas de poetry y las ventajas de pip y descartar sus falencias. Revisa el archivo `pyproject.toml` para agregar las librerias, usa el siguiente link para entender como agregar la [versión](https://python-poetry.org/docs/dependency-specification/).

### Manejo de branch

Uso git `hubflow` para el trabajo con las ramas
Para iniciar el uso de hubflow en el proyecto por primera vez usa `git hf init`
Para crear una nueva rama `git hf TAG start NOMBRE-DE-LA-RAMA` recuerda que TAG varia dependiendo del objetivo de la rama.
Debes mezclar la rama según su flujo normal `git push origin TAG/NOMBRE-DE-LA-RAMA`.
Para actualizar la rama y `develop` con los cambios que esten arriba ejecuta `git hf update`.
Elimina la rama localmente y en github `git hf feature finish nombre-de-la-rama`.

### Testing

Se pueden ejecutar las pruebas de python usando el comando:
```sh
./cmd -tp | --tests_python
```

Se pueden ejecutar las pruebas de node usando el comando:
```sh
./cmd -tn | --tests_npm
```

### Deploy de una nueva versión.

Para generar una nueva versión, se debe crear un PR (con un título "Release X.Y.Z" con los valores que correspondan para X, Y y Z). Se debe seguir el estándar semver para determinar si se incrementa el valor de X (si hay cambios no retrocompatibles), Y (para mejoras retrocompatibles) o Z (si sólo hubo correcciones a bugs).

En ese PR deben incluirse los siguientes cambios:

Modificar el archivo `CHANGELOG.md` y `VERSION` para incluir una nueva entrada (al comienzo) para X.Y.Z que explique los cambios.
Modificar el archivo `poetry.toml` para que la propiedad `"version"` apunte a la nueva versión X.Y.Z

Luego de obtener aprobación del pull request, debe mezclarse a master e inmediatamente generar un release en GitHub con el tag X.Y.Z. En la descripción del release debes poner lo mismo que agregaste al changelog.


## Paso a producción

Aún no esta listo para producción, se deben cambiar manualmente el `hosts: production` en cada archivo del contenedor
de ansible config-(contenedor).yml

```sh
docker-compose -f docker-compose.yml -f docker-compose.prod.yml up --build
```


## Notas

Para validar que los servicios estan arriba al usar docker
```sh
nmap 0.0.0.0 -p 23,24,25,80,5432,5555,7473,7474,7687,8001 | grep -i tcp
```

Cada contenedor contiene una llave para el Vault de ansible, debe ingresar a cada uno y revisar
el archivo crea un archivo .key dentro de la carpeta ansible, si desea cambiar las contraseñas desencrypta los valores y vuelve a encryptar con una nueva llave.
En la configuración de ansible puede buscar la palabra ".key" y comentar la linea para que no busque
la contraseña en un archivo y así asignar una contraseña nueva y guardarla en algún lugar seguro. Esto puede tener efectos en la construcción del contenedor.


### Usar y obtener actualziaciones desde la plantilla.

En gitHub crea un repositorio desde el template de `djfullapp`, para poder obtener los cambios que se generen en el template se deben realizar varias acciones que son importantes comprender.

#### Preparar rama para recibir actualizaciones

- Clona `djfullapp` y desde la rama `master` crea una rama con el mismo nombre del nuevo repositorio (por razones practicas) en este caso será `blog`, esta rama no se empujará a `GIT` así que no será visible en el repositorio de `djfullapp`, tu repositorio privado permanece privado.

```sh
git fetch origin
git checkout master
git branch -b blog
```

#### Control de versiones en mi nuevo repositorio

- Se creo un proyecto usando el template, para este ejemplo es `blog` (desde github).
- Además agrego un nuevo repositorio remoto a `djfullapp` que apunta a ese nuevo repositorio.

```sh
git remote add blog https://github.com/saengate/blog.git
```
 
`Nota intermedia`: Aquí se pone interesante, usando las propiedades de git las actualizaciones de `djfullapp` que quiero empujar en `blog` los realizo dentro de esta rama, por lo que esta rama será la branch intermedio de estos proyectos.


- `Ten cuidado aquí`: Y empuja la rama `blog` dentro del nuevo proyecto cambiando origin por blog:

```sh
git push blog blog
```

- En Github asigna esta rama como la rama principal del proyecto, recuerda que debes hacerlo en el settings de ese repositorio github.

#### Actualizar BLOG con los nuevas versiones de DJFULLAPP

Desde el proyecto `djfullapp` los cambios hacia esta nueva rama se empujan con los siguientes comandos:

Ten cuidado con los comandos que ejecutes aquí, un mal paso puede reescribir el repositorio equivocado.
También recuerda que origin apunta a `djfullapp` y `blog`, bueno a `blog`.
- Cambia a la rama intermediaria `git checkout blog`.
- Actualiza rama intermedia con los cambios que hayas realizado `git pull blog blog`.
- Actualiza la rama `master` de `djfullapp`: `git pull origin master`.
- Mezcla los cambios de `djfullapp` en `blog`:
```sh
git checkout blog
git rebase master
git push blog blog
```
Tendrás que lidiar con algunos conflictos, pero no deberían ser demasiados. Tal vez todos asociados a la carpeta `ses`.
No borres la rama intermedia, `blog` para este caso.

#### Enviar README.md a dockerhub

```sh
export DOCKERHUB_USERNAME="usuario";
export DOCKERHUB_PASSWORD="contraseña";
export DOCKERHUB_REPO_PREFIX="usuario_o_prefijo_en_docker_hub";
export DOCKERHUB_REPO_NAME="repositorio";
docker run --rm -v "$(pwd)/:/data/:ro" -e "DOCKERHUB_USERNAME=$DOCKERHUB_USERNAME" -e "DOCKERHUB_PASSWORD=$DOCKERHUB_PASSWORD" -e "DOCKERHUB_REPO_PREFIX=$DOCKERHUB_REPO_PREFIX" -e "DOCKERHUB_REPO_NAME=$DOCKERHUB_REPO_NAME" sheogorath/readme-to-dockerhub;
```

## Errores conocidos

### Supervisor

```sh
unix:///var/run/supervisor.sock no such file
```

En ocasiones necesitaras reiniciar el servicio y verás este error, el problema se debe a que hace falta cargar
`cada vez` las configuraciones de supervisor, el comando es simple pero para simplificar la tarea ejecuta el siguiente
comando que carga las configuraciones y refresta todo el entorno supervisor. (Si quieres conocer los detalles,
revisa el archivo en ansible template o dentro del projecto o en /usr/local/bin/runsupervisor).

```sh
runsupervisor
```


# Despliegue en producción

* URL: `www.saengate.com`

Usa AWS para su despliegue.

* Región: EE.UU. Este (Norte de Virginia) us-east-1

TODO: Cambiar a `Amazon Aurora Serverless: PostgreSQL v10.12`
## Configuración de la base de datos (Modo prueba)

Tomar en consideración para reducir al minimo posibles costos:
1. Elegir la capa gratuita.
2. Elegir en todas las secciones el recurso más bajo.
3. No usar backups.

* `DB instance identifier`: wagtail
* `Instance`: db.t2.micro

* Debe crearse una nueva VPC y configurar el puerto de acceso junto con la ruta de tablas.
TODO: documentar


CHANGELOG.md
# Changelog
Todos los cambios notables a este proyecto serán documentados en este archivo.

El formato está basado en [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
y este proyecto adhiere a [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.0.0] - YYYY-MM-DD
### Added
### Fixed
### Feature
- [Descripción.](https://github.com/saengate/djfullapp/pull/#)

## [1.0.0] - 2021-01-07
### Feature
- [Primera versión del blog saengate con wagtail en AWS.](https://github.com/saengate/wagtail-compose/pull/1)


project/config/settings.py
"""
Django settings for mysite project.

Generated by 'django-admin startproject' using Django 2.2.7.

For more information on this file, see
https://docs.djangoproject.com/en/2.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/2.2/ref/settings/
"""

import os

from os import getenv
from rest_framework import compat

from utils.color_logging import formatter

# I have the same issue, but poetry does not allow me to update Markdown to from 2.6.11 to 3+
# because apache-airflow (1.10.10) depends on markdown (>=2.5.2,<3.0)
compat.md_filter_add_syntax_highlight = lambda md: False

ENVIRONMENT = 'develop'

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PROJECT_DIR = os.path.abspath(BASE_DIR)


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/2.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = 'e7(sinr28hr10@9(+765wd&*0jjl%_qj1sxckjw!ssn^x0dm7@'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = ['*']


# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'rest_framework',
    'channels',
    'django_neomodel',
    'easyaudit',
    'ses',
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'easyaudit.middleware.easyaudit.EasyAuditMiddleware',
]

ROOT_URLCONF = 'config.urls'


REST_FRAMEWORK = {
    # Use Django's standard `django.contrib.auth` permissions,
    # or allow read-only access for unauthenticated users.
    'DEFAULT_PERMISSION_CLASSES': [
        'rest_framework.permissions.AllowAny',
        # 'rest_framework.permissions.DjangoModelPermissions',
        # 'rest_framework.permissions.DjangoModelPermissionsOrAnonReadOnly,
    ],
    'DEFAULT_RENDERER_CLASSES': [
        'rest_framework.renderers.JSONRenderer',
        'rest_framework.renderers.BrowsableAPIRenderer',
    ],
}


TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'django.template.context_processors.i18n',
            ],
        },
    },
]

WSGI_APPLICATION = 'config.wsgi.application'


# Database
# https://docs.djangoproject.com/en/2.2/ref/settings/#databases

"""
'default': {
    'ENGINE': 'django.db.backends.sqlite3',
    'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
}
"""
# https://pypi.org/project/django_neomodel/

NEOMODEL_NEO4J_BOLT_URL = 'bolt://neo4j:neo4j@djfullapp-neo4j:7687'
NEOMODEL_SIGNALS = True
NEOMODEL_FORCE_TIMEZONE = False
NEOMODEL_ENCRYPTED_CONNECTION = True
NEOMODEL_MAX_POOL_SIZE = 50

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'djfullappdb',
        'USER': 'userdb',
        'PASSWORD': 'password',
        'HOST': 'djfullapp-db',
        'PORT': '5432',
    }
}


# Password validation
# https://docs.djangoproject.com/en/2.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.2/howto/static-files/

STATICFILES_FINDERS = [
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
]

STATICFILES_DIRS = [
    os.path.join(PROJECT_DIR, 'staticfiles'),
]

STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'static')

MEDIA_URL = '/media/'
MEDIA_ROOT = os.path.join(BASE_DIR, 'media')


# Internationalization
# https://docs.djangoproject.com/en/2.2/topics/i18n/

# django-admin makemessages -a && django-admin compilemessages

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

LOCALE_PATHS = [
    os.path.join(BASE_DIR, 'locale'),
]

LANGUAGE_CODE = 'es-cl'
# LANGUAGE_CODE = 'en-us'

_ = lambda s: s  # NOQA

LANGUAGES = [
    ('es', _('Espanish')),
    ('en', _('English')),
]


# ==============================================================================
# WEBSOCKETS & CHANNELS
# ==============================================================================

ASGI_APPLICATION = 'config.routing.application'

REDIS_HOST = getenv('REDIS_HOST', 'redis://redis:6379')

# It is possible to have multiple channel layers configured.
CHANNEL_LAYERS = {
    'default': {
        'BACKEND': 'channels_redis.core.RedisChannelLayer',
        'CONFIG': {
            "hosts": [REDIS_HOST],
        },
    },
}

HOSTS = {
    'production': ['127.0.0.1'],
    'develop': ['127.0.0.1'],
}

WS_ALLOWED_HOSTS = ['*'] if DEBUG else HOSTS[ENVIRONMENT]


# LOGS
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'standard': {
            'format': formatter(),
        },
    },
    'handlers': {
        'file': {
            'level': 'INFO',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': '/var/log/project/project.log',
            'maxBytes': 1024*1024*10,  # 10MB
            'backupCount': 5,
            'formatter': 'standard',
        },
        'console': {
            'class': 'logging.StreamHandler',
            'formatter': 'standard',
        },
    },
    'loggers': {
        '': {
            'class': 'utils.color_logging.NewLogger',
            'handlers': ['file'],
            'level': 'DEBUG',
            'propagate': True,
        },
        'console': {
            'class': 'utils.color_logging.NewLogger',
            'handlers': ['console'],
            'level': 'DEBUG',
            'propagate': True,
        },
        'all': {
            'class': 'utils.color_logging.NewLogger',
            'handlers': ['file', 'console'],
            'level': 'DEBUG',
            'propagate': True,
        },
    },
}


# AUDIT [django-easy-audit](https://github.com/soynatan/django-easy-audit)

DJANGO_EASY_AUDIT_WATCH_MODEL_EVENTS = True
DJANGO_EASY_AUDIT_WATCH_AUTH_EVENTS = True
DJANGO_EASY_AUDIT_WATCH_REQUEST_EVENTS = True

# DJANGO_EASY_AUDIT_UNREGISTERED_CLASSES_EXTRA = ['app_name.model_name']
# DJANGO_EASY_AUDIT_UNREGISTERED_URLS_EXTRA = ['url']
# DJANGO_EASY_AUDIT_CRUD_DIFFERENCE_CALLBACKS = ['string-paths-to-functions-classes']

# This is reserved for future use (does not do anything yet) 1.2.2
# DJANGO_EASY_AUDIT_USER_DB_CONSTRAINT = 'default'

# DJANGO_EASY_AUDIT_CRUD_EVENT_LIST_FILTER = ['event_type', 'content_type', 'user', 'datetime', ] 
# DJANGO_EASY_AUDIT_LOGIN_EVENT_LIST_FILTER = ['login_type', 'user', 'datetime', ]
# DJANGO_EASY_AUDIT_REQUEST_EVENT_LIST_FILTER = ['method', 'user', 'datetime', ]

# DJANGO_EASY_AUDIT_DATABASE_ALIAS = 'default '

# No guarda auditoria si no han ocurrido cambios
DJANGO_EASY_AUDIT_CRUD_EVENT_NO_CHANGED_FIELDS_SKIP = True

# Solo lectura impide que un superusuario los modifique
DJANGO_EASY_AUDIT_READONLY_EVENTS = True

# DJANGO_EASY_AUDIT_LOGGING_BACKEND = 'easyaudit.backends.ModelBackend'
